### ISSUE TITLE
Mac OS: bad quality of drawing

### ISSUE THREAD

**Issue by rcaneill** [xournalpp/xournalpp#5451]
### Operating System

Mac OS

### (Linux Only) Display Server

None

### Installation Method

Official Release

### Xournal++ Version

1.2.2

### libgtk Version

3.24.38

### Bug Description

Hi,

First, thanks for this great software :).

On the Mac (version sonoma 14.2.1 (23C71)), the drawings are very laggy and inaccurate.
This happens despite using a good quality Wacom graphic tablet (which works fine with other softwares, as e.g illustrator).

![photo1705160567](https://github.com/xournalpp/xournalpp/assets/18579092/a0194135-95db-4d0d-b9dd-4aca62607bc1)

### Expected Behaviour

The expected behaviour is that the lines are smooth and accurate.

### Steps to Reproduce

1. Open Xournal++
2. draw anything

**Comment by Febbe** [comment]
@rolandlo can you reproduce this behavior on your Mac?
@rcaneill there is a workaround for the "too few input events" problem, whichever the cause is (slow device, high workload etc.).
You can activate to interpolate + smooth all events. That should increase the quality, but it adds a small delay to drawing, since we must buffer some events for the spline & flattening filter.

**Comment by rcaneill** [comment]
Thanks for the answer! We'll try the workaround and see how it goes.
Would still be nice to get it to work natively, though.

**Comment by uzluisf** [comment]
Related: https://github.com/xournalpp/xournalpp/issues/5530

**Comment by E8D08F** [comment]
I tried Xournal++ on macOS with a Wacom tablet via Type-C cable and can confirm the issue. The points that connect the strokes seem to be attached to _integer_ pixels on the screen, which makes the handwriting zigzag if drawing an oblique line. Connecting to the same Wacom device, I tested on both Ubuntu virtual machine and Steam Deck, they produced smoother curves with the default settings. Can we safely conclude that the behaviour of GTK’s cursor on macOS event is the cause of this intrinsic inferiority?

I also tested the native macOS AppKit. Mouse events (down/up, drag) provide floating-point positions of the pen tip. I believe this makes it easier to construct a natural writing experience.

**Comment by rolandlo** [comment]
> I tried Xournal++ on macOS with a Wacom tablet via Type-C cable and can confirm the issue. The points that connect the strokes seem to be attached to _integer_ pixels on the screen, which makes the handwriting zigzag if drawing an oblique line.

Great observation! That explains a lot. I can confirm. On MacOS event x- and y- coordinates (as shown in the logs from the testing area, both in the `full-logs.txt` and `input-gdk_events.txt`) are all integers. This happens for both stylus and mouse. I even use the same Logitech mouse and Wacom tablet like on Linux, where coordinates are true floating-point.

Pressure values are floating-point on MacOS as well.

**Comment by E8D08F** [comment]
May I ask wether this can be fixed within Xournal++, or it needs to be addressed on the GTK side?

**Comment by rolandlo** [comment]
Looking at [gdk-events-quartz.c](https://gitlab.gnome.org/GNOME/gtk/-/blob/3.24.51/gdk/quartz/gdkevents-quartz.c?ref_type=tags#L529)

it's no wonder: `motion.x`, `motion.y`, `motion.x_root` and `motion.y_root` are `gint`s.

```C
static void
generate_motion_event (GdkWindow *window)
{
  NSPoint screen_point;
  GdkEvent *event;
  gint x, y, x_root, y_root;
  GdkDisplay *display = gdk_window_get_display (window);
  GdkSeat *seat = gdk_display_get_default_seat (display);

  event = gdk_event_new (GDK_MOTION_NOTIFY);
  event->any.window = NULL;
  event->any.send_event = TRUE;

  screen_point = [NSEvent mouseLocation];

  _gdk_quartz_window_nspoint_to_gdk_xy (screen_point, &x_root, &y_root);
  get_window_point_from_screen_point (window, screen_point, &x, &y);

  event->any.type = GDK_MOTION_NOTIFY;
  event->motion.window = window;
  event->motion.time = get_time_from_ns_event ([NSApp currentEvent]);
  event->motion.x = x;
  event->motion.y = y;
  event->motion.x_root = x_root;
  event->motion.y_root = y_root;
```

The rounding happens on a lower level in [gdkscreen-quartz.c](https://gitlab.gnome.org/GNOME/gtk/-/blob/3.24.51/gdk/quartz/gdkscreen-quartz.c#L121):
```C
static void
gdk_quartz_screen_calculate_layout (GdkQuartzScreen *screen,
                                    GdkQuartzDisplay *display)
{
  if (!display)
    display = GDK_QUARTZ_DISPLAY (gdk_screen_get_display (GDK_SCREEN (screen)));

/* Display geometry is the origin and size in AppKit coordinates. AppKit computes */
  screen->width = (int)trunc (display->geometry.size.width);
  screen->height = (int)trunc (display->geometry.size.height);
  screen->orig_x = -(int)trunc (display->geometry.origin.x);
  screen->orig_y = (int)trunc (display->geometry.origin.y);
  screen->mm_width = (int)trunc (display->size.width);
  screen->mm_height = (int)trunc (display->size.height);

 }
```
Here [GdkQuartzDisplay](https://gitlab.gnome.org/GNOME/gtk/-/blob/3.24.51/gdk/quartz/gdkdisplay-quartz.h#L34) still uses floating-point coordinates:

```C
struct _GdkQuartzDisplay
{
  GdkDisplay parent_instance;
  NSRect geometry; /* In AppKit coordinates. */
  NSSize size; /* Aggregate size of displays in millimeters. */
  GPtrArray *monitors;
  /* This structure is not allocated. It points to an embedded
   * GList in the GdkWindow. */
  GSList   *windows_awaiting_frame;
  GSource *frame_source;
};
```
See also the documentation for [NSRect](https://developer.apple.com/documentation/Foundation/NSRect)

**Comment by rolandlo** [comment]
> May I ask wether this can be fixed within Xournal++, or it needs to be addressed on the GTK side?

This must definitely be addressed on the GDK side. Even if it is addressed there, they won't put out a new release until March 2027 as they recently announced (see https://gitlab.gnome.org/GNOME/gtk/-/blob/gtk-3-24/NEWS)
I'm wondering if Gdk 4 still uses integer coordinates on MacOS.

**Comment by rolandlo** [comment]
> > May I ask wether this can be fixed within Xournal++, or it needs to be addressed on the GTK side?
> 
> This must definitely be addressed on the GDK side. Even if it is addressed there, they won't put out a new release until March 2027 as they recently announced (see https://gitlab.gnome.org/GNOME/gtk/-/blob/gtk-3-24/NEWS) I'm wondering if Gdk 4 still uses integer coordinates on MacOS.

Yes, Gdk 4 also uses integer coordinates on MacOS. See the file [gdkmacosdisplay-translate.c](https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gdk/macos/gdkmacosdisplay-translate.c#L564). I haven't found an issue on the Gtk issue tracker, so I have opened one: https://gitlab.gnome.org/GNOME/gtk/-/issues/8132

**Comment by andylovephysics** [comment]
I recently tried an alternative method to install xournalpp on macOS 12, which seems to give a better quality of drawing.

Instead of installing directly the dmg package in xournalpp release, I tried to compile the xorg based xournalpp. 

> System Info: macos12 macbook pro 2015

I installed macport 2.12.5, wacom tablet driver, xquartz2.8.6 first. 
Using `port install xournalpp`, I accidentally installed the xournalpp that can work on xquartz. The default version for xournalpp is 1.2.3. After xournalpp has been installed, just run `xournalpp` in terminal, and get a drawable xournalpp.

How drawing looks like:

<img width="194" height="159" alt="Image" src="https://github.com/user-attachments/assets/3391cbd6-d3e5-45cb-8fe1-97603af51940" />


The problem I noticed are:
 - bad resolution, both GUI and the content on canvas
 - elements in a selected group after being pasted are no longer usable. (Can be solved by toggling off: Prefrecnes/Pasteboard/Update pastebard when clipboard changes)
